### ISSUE TITLE
Typing - letter rendering lagging

### ISSUE THREAD

**Issue by Lledor** [vkbo/novelWriter#2129]
Hi,

I’m currently using Version 2.6 Beta 1 on Windows 10 and have noticed occasional lags while typing. There’s a noticeable delay between typing and the text appearing on the screen. This issue is specific to this application, as other programs on my computer are functioning normally.

Thank you for your assistance.

**Comment by Lledor** [comment]
Okay, I follow your recommendation - thank you for your help.

**Comment by vkbo** [comment]
Did changing any settings make a difference?

**Comment by Lledor** [comment]
Yes, in my case the auto-replace, was the culprit. Apologies for not letting you know earlier.

**Comment by vkbo** [comment]
Sure, no problem. The auto-replacer keeps growing slowly. and there's quite a lot of code to run through on each keypress. I'll look into optimising it a little.

**Comment by vkbo** [comment]
Hmm, with English settings, it takes about 30 microsoeconds to process on my laptop. That shouldn't be noticeable. What kind of auto-replace settings did you have? Any of the special French/Spanish ones that inserts spaces around punctuation?

**Comment by Lledor** [comment]
UK English only, spell check on and nothing much else. Syntax highlighting on did not affect it, only the auto-replace. It may even be my small desktop computer? Windows 10 System Type: x64-based PC Processor:Intel(R) Celeron(R) J4125 CPU @ 2.00GHz, 2001 Mhz, 4 Core(s), 4 Logical Processor(s). I am in the process of moving my work over from Scrivener, but that is idle as I work in novelWriter.

**Comment by vkbo** [comment]
It could just be processing priority or something like that. novelWriter is designed to be light on resources, but Python is also fairly slow, so anything that requires real-time processing as you type can potentially be an issue. Since the problem is intermittent, I expect resources do play a role here.

I also get the impression the whole framework novelWriter runs on is a little slower on Windows than on Linux. So it may very well be that it's not a matter of how fast the auto-replace feature can run, but the fact that it has to run in the same thread as the text rendering at all. That's a Qt requirement, and unavoidable. So optimising the code might not help, as the problem may be that this part of the code has to run at all. If that makes sense.

Do these intermittent delays coincide with either auto-save (every 30 seconds by default) or word counting (every 5 seconds by default)?

**Comment by Lledor** [comment]
The settings are default, save document 30 seconds, word counting 5 seconds. I will moving over to Linux Mint in the near future, this is my reason for dumping Scrivener (apart from it being too bloated an application) so it is good to know that it work better on Linux. I think that move may be quicker that I anticipated?

**Comment by vkbo** [comment]
I tested this on a Windows virtual machine, so a fairly low resource environment, and I can reproduce your issue. However, I cannot make it go away by turning off auto-replace. I suspect this has nothing to do with the auto-replace itself, but a rendering delay. There were minimal resource usage, and the auto-replace took nowhere enough time to cause the delay. Perhaps turning it off in your case keeps it just below the threshold where it starts lagging.

A little web search revealed similar reports on Windows, even on minimal example code, so this may be Qt-related. It's hard to say. I would like to hold off on these kind of potential Qt-related issues until novelWriter moves to Qt6, which may or may not fix some of them. It will be a good jump forward in versions.

The plan is to make the switch in 2.7, ahead of the Qt5 end of life in May.

**Comment by vkbo** [comment]
Ok, got another bug report on typing lag. Are you still seeing this issue? if so, was this issue new for you as well for 2.6.x, and not present in 2.5.x?

I've looked through the code changes from 2.5.3 to 2.6 Beta 2, and there is a lot, but nothing that I could see at first look-through. Some of the code changed can potentially affect the typing, but there are also other things that can. I'm trying to narrow down if there is a specific version that caused this to get slower.

**Comment by xahodo** [comment]
Just to add to the discussion:

- I live on Linux Mint 22.1
- Running NovelWriter 2.7b1
- Hardware: Ryzen 3 4300U at 2.7GHz, 8GB RAM

Not seeing the issue. Then again: my current largest document is 545 words, so there's not that much work to do for NW.

Perhaps the issue relates to the size of the document (as seen by Qt)?

**Comment by vkbo** [comment]
It shouldn't. Qt is specifically optimised for large documents. Only the line you're editing is updating. That's what the syntax highlighter and spell checker runs on too.

If you have a very long paragraph, you will eventually run into lag of course. But any normal size paragraph should be fine.

There are a few large document issues, but those are related to full document updates, like turning on and off the spell checker.

**Comment by johnblommers** [comment]
> Not seeing the issue.

I'm on PopOS 22.04 with a roughly 50,000 word manuscript.

The novelWriter 2.7b1 AppImage works fine here.

**Comment by vkbo** [comment]
I suspect the issue is related to graphics acceleration and how quickly the Qt library can respond to updates. That would explain why even the most basic app can sometimes lag.

novelWriter processes a few things on each keystroke, and as I mentioned above, if the process time including rendering exceeds the average time between keystrokes, a lag will build up because Qt processes the editor updates sequentially in the GUI main thread. (It is very hard to process any document content in the background, which is an issue for stuff like spell checking I have not yet solved.)

There's a chance the 2.7 release will work better because it runs on the very latest version of Qt6, while 2.6 and earlier ran on a much older version of Qt5 that is reaching end of life next month. They haven't been fixing bugs there for years.

**Comment by vkbo** [comment]
I just moved this to the 2.7.1 milestone. I will release 2.7 tomorrow on 1st June, so you should be able to test if the 2.7 release fixes it. If I don't hear anything else, I will close this issue ticket after that. As mentioned, since I have not been able to reproduce the problem, I cannot on my own verify if the problem still exists.

**Comment by vkbo** [comment]
Closing this due to lack of feedback.
